## Module 6 Project
By: Ken Wilson

Wide World Importers is a fake company that sells a variety of different products such as slippers, coffee mugs, and flash drives. The company has extensive records of invoices that have been created during the past several years and would like you, the data analyst, to explore the data to determine if there is anything noteworthy that executives should incorporate into their business strategy. However, because the data has been input over the years using several different programs, there are some holes in the data set that need "cleaning". The data will eventually be run through a machine learning model and thus cannot contain any null values.

Download the data set [invoices.csv](https://btech.instructure.com/courses/610804/files/118105343/download?download_frd=1)

Download invoices.csv by clicking the link and follow the steps below to clean up the data set. Set up a new Google Colab file as an exploratory data analysis and use the file to clean the data. At the end of this assignment, you will turn in the Google Colab file.

You should perform the following cleaning steps as indicated in the questions below, in this order:

    Fix standardization errors, where applicable
    Fix business logic errors, where applicable
    Impute null values, where possible
    Drop columns with many null values, where applicable
    Drop rows with null values, where applicable
    Drop duplicate rows, where applicable
    Identify outliers and fix them as instructed

The company would like you to answer the following questions, after cleaning has been performed:

    Which customer has the most invoices?
    Which item has the highest total quantity sold?
    Which item has generated the most total profit?
    Which item do customers purchase the most quantity of, on average?




In [1]:
import pandas as pd
df = pd.read_csv('invoices.csv', low_memory=False)
display(df.head())
display(df.info())

InvoiceLineID  InvoiceID  StockItemID  \
0              1          1         67.0   
1              2          2         50.0   
2              3          2         10.0   
3              4          3        114.0   
4              5          4        206.0   

                                         Description  PackageTypeID  Quantity  \
0           Ride on toy sedan car (Black) 1/12 scale              7        10   
1  Developer joke mug - old C developers never di...              7         9   
2               USB food flash drive - chocolate bar              7         9   
3                 Superhero action jacket (Blue) XXL              7         3   
4         Permanent marker black 5mm nib (Black) 5mm              7        96   

   UnitPrice  TaxRate  TaxAmount  LineProfit  ExtendedPrice  LastEditedBy  \
0      230.0       15      345.0       850.0         2645.0             7   
1       13.0       15        NaN        76.5         -999.0             7   
2       32.0       15       43.2       180.0          331.2             7   
3       30.0       15        NaN        24.0         -999.0             7   
4        2.7       15        NaN        96.0            NaN             7   

   LastEditedWhen          Customer  
0             NaN     Stuff by Stew  
1  1/1/2013 12:00   Wholesaler Plus  
2             NaN   Wholesaler Plus  
3             NaN   Big Buys Retail  
4  1/1/2013 12:00  Terry's Trinkets

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 230548 entries, 0 to 230547
Data columns (total 14 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   InvoiceLineID   230548 non-null  int64  
 1   InvoiceID       230548 non-null  int64  
 2   StockItemID     230087 non-null  float64
 3   Description     229876 non-null  object 
 4   PackageTypeID   230548 non-null  int64  
 5   Quantity        230548 non-null  int64  
 6   UnitPrice       230548 non-null  float64
 7   TaxRate         230548 non-null  int64  
 8   TaxAmount       109623 non-null  float64
 9   LineProfit      230548 non-null  float64
 10  ExtendedPrice   113081 non-null  float64
 11  LastEditedBy    230548 non-null  int64  
 12  LastEditedWhen  6 non-null       object 
 13  Customer        230548 non-null  object 
dtypes: float64(5), int64(6), object(3)
memory usage: 24.6+ MB


None

### **Question 1**
How many rows does the data set have?

Group of answer choices
* 115,484
* 189,773
* 230,548
* 98,707
### Answer: 230,548

In [2]:
df.shape[0]

230548

### **Question 2**
Which of the following columns contain null values?

Group of answer choices
* LastEditedWhen
* UnitPrice
* StockItemID
* Customer
### Answer: LastEditedWhen, StockItemID

In [3]:
df.isna().sum().reset_index().sort_values(by = 0, ascending = False).reset_index(drop = True).rename(columns={'index' : 'columns', 0 : 'count_of_Nulls'})

columns  count_of_Nulls
0   LastEditedWhen          230542
1        TaxAmount          120925
2    ExtendedPrice          117467
3      Description             672
4      StockItemID             461
5        InvoiceID               0
6         Quantity               0
7    PackageTypeID               0
8    InvoiceLineID               0
9        UnitPrice               0
10      LineProfit               0
11         TaxRate               0
12    LastEditedBy               0
13        Customer               0

### **Question 3**
Compute the descriptive statistics on the data set. Which columns appear to have logical errors?

Group of answer choices
* the highest quantity is more than 3,000 which is impossible
* the lowest extended price is a negative number
* the lowest Tax Rate is a negative number
* the lowest quantity is a negative number
### Answers:
* the lowest extended price is a negative number
* the lowest Tax Rate is a negative number

In [4]:
df.describe()

InvoiceLineID      InvoiceID    StockItemID  PackageTypeID  \
count  230548.000000  230548.000000  230087.000000  230548.000000   
mean   114127.066880   35177.386891     110.181827       7.073638   
std     65900.142929   20338.920057      63.722514       0.644528   
min         1.000000       1.000000       1.000000       1.000000   
25%     57062.750000   17570.000000      54.000000       7.000000   
50%    114130.500000   35151.000000     111.000000       7.000000   
75%    171203.250000   52766.000000     165.000000       7.000000   
max    228265.000000   70510.000000     227.000000      10.000000   

            Quantity      UnitPrice        TaxRate      TaxAmount  \
count  230548.000000  230548.000000  230548.000000  109623.000000   
mean       39.252524      45.539361      14.463409     113.328006   
std        55.193934     139.510031       3.904056     221.866696   
min         1.000000       0.660000     -15.000000       0.380000   
25%         5.000000      13.000000      15.000000      14.400000   
50%        10.000000      18.000000      15.000000      34.500000   
75%        60.000000      32.000000      15.000000     129.600000   
max      3150.000000    1899.000000      15.000000   14994.000000   

         LineProfit  ExtendedPrice   LastEditedBy  
count  230548.00000  113081.000000  230548.000000  
mean      375.38894     740.092626      10.800480  
std       753.17041    1705.229089       5.509623  
min      -645.00000    -999.000000       2.000000  
25%        51.00000      89.700000       6.000000  
50%       120.00000     248.400000      11.000000  
75%       390.00000     919.080000      16.000000  
max      9200.00000  114954.000000      20.000000

### **Question 4**
Why might a negative number in the LineProfit column not be considered a logic error?

Group of answer choices
* a negative line profit indicates that the profit was computed incorrectly
* a negative profit is not possible
* a product could, in theory, be sold at a loss
* the invoice probably comes from a different data set
### Answer: a product could, in theory, be sold at a loss

### **Question 5**
Fix standardization issues in the data set to determine which customer has created the most unique invoices. Which customer created the most invoices?

Group of answer choices
* Wholesaler Plus
* Terry's Trinkets
* Stuff by Stew
* Big Buys Retail
### Answer: Stuff by Stew

In [5]:
print('\nBefore:\n')
display(df['Customer'].value_counts())

### Do the cleaning
df['Customer'] = df['Customer'].str.title()
df.replace({'Customer' : {'Stuf By Stew' : 'Stuff By Stew', "Terry'S Trinket" : "Terry's Trinkets", "Terry'S Trinkets" : "Terry's Trinkets"}}, inplace = True)

print('\n\nAfter:\n')
display(df['Customer'].value_counts())


Before:



Customer
Terry's Trinkets       85593
Stuff by Stew          62752
Stuf by Stew           31984
Wholesaler Plus        15735
Big Buys Retail        13461
Terry's Trinket         9483
stuff by stew           7068
Dennis Distribution     4472
Name: count, dtype: int64



After:



Customer
Stuff By Stew          101804
Terry's Trinkets        95076
Wholesaler Plus         15735
Big Buys Retail         13461
Dennis Distribution      4472
Name: count, dtype: int64

### **Question 6**
Observe the null values in the LastEditedWhen column. Based on the number of null values in the column and the questions being asked by executives, what do you recommend doing to eliminate null values?

Group of answer choices
* drop the column
* do nothing
* drop the rows with null values in that column
* use other information in the data set to impute the dates
### Answer: drop the column
* The 'LastEditedWhen' values will not have any impact on this study.  Let's just drop it.

### **Question 7**
Perform the action that you recommended above. After performing the action, which column contains the most null values?

Group of answer choices
* StockItemID
* Description
* TaxAmount
* ExtendedPrice
### Answer: TaxAmount

In [6]:
df.drop(columns = 'LastEditedWhen', inplace = True)
df.isna().sum().reset_index().sort_values(by = 0, ascending = False).reset_index(drop = True).rename(columns={'index' : 'columns', 0 : 'count_of_Nulls'})

columns  count_of_Nulls
0       TaxAmount          120925
1   ExtendedPrice          117467
2     Description             672
3     StockItemID             461
4   InvoiceLineID               0
5   PackageTypeID               0
6       InvoiceID               0
7       UnitPrice               0
8        Quantity               0
9         TaxRate               0
10     LineProfit               0
11   LastEditedBy               0
12       Customer               0

### **Question 8**
Observe the columns StockItemID and Description. Notice that StockItemID is a single number, and each number has a corresponding Description. Thus, by knowing the StockItemID of a row, its Description can be imputed. However, if both the StockItemID and Description are missing, the item is unknown.

How many rows have both a null StockItemID and a null Description?

Group of answer choices
* 102
* 554
* 461
* 306
### Answer: 461

In [7]:
print('Number of rows having StockItemId and Description both Null:')
print(df.shape[0] - len(df.dropna(subset = ['StockItemID', 'Description'], how = 'all')))

# Or...
#len(df[(df['Description'].isna()) & (df['StockItemID'].isna())])

Number of rows having StockItemId and Description both Null:
461


### **Question 9**
Observe the rows that have both a null StockItemID and a null Description. Based on the number of null values and the questions asked by the executives, what do you recommend doing to remove null values from these columns?

Group of answer choices
* do nothing
* drop the rows with null values in the Description and StockItemID columns
* impute the missing values using information in the data set
* drop the StockItemID and Description columns
### Answer: <u>drop the rows with null values in the Description and StockItemID columns</u>
  * The executives want item numbers & descriptions of those items that have the highest quantity sold, most total profit, and most quantity purchase average. We won't have the identity of these NaN items, so they won't be helpful in this case.
  * Maybe more importanly, a requirement for the final output is that: "The data will eventually be run through a machine learning model and thus cannot contain any null values."

  An argument for keeping the rows, maybe?
  * The executives want item numbers & descriptions of those items that have the highest quantity sold, most total profit, and most quantity purchase average. We won't have the identity of these NaN items, but there is still valid information on these sales.
  * If any of these 'items' end up being the ones that meet the 'highest/most' criteria or appear significant, the executives would want to know and they would want additional research done on those missing values.

In [8]:
df[(df['Description'].isna()) & (df['StockItemID'].isna())].describe()

InvoiceLineID     InvoiceID  StockItemID  PackageTypeID    Quantity  \
count     461.000000    461.000000          0.0     461.000000  461.000000   
mean   112079.548807  34549.648590          NaN       7.158351   41.876356   
std     68765.368575  21224.946465          NaN       0.703161   55.541797   
min       597.000000    251.000000          NaN       1.000000    1.000000   
25%     51382.000000  15839.000000          NaN       7.000000    5.000000   
50%    113336.000000  34901.000000          NaN       7.000000   10.000000   
75%    174077.000000  53655.000000          NaN       7.000000   60.000000   
max    227910.000000  70407.000000          NaN      10.000000  252.000000   

        UnitPrice     TaxRate    TaxAmount   LineProfit  ExtendedPrice  \
count   461.00000  461.000000   215.000000   461.000000     223.000000   
mean     48.88692   14.533623   130.333209   412.370716     864.816996   
std     162.09547    3.678589   255.183145   800.544226    1991.010590   
min       0.66000  -15.000000     1.950000  -120.000000    -999.000000   
25%      13.00000   15.000000    15.600000    50.000000      92.720000   
50%      18.00000   15.000000    36.000000   135.000000     257.600000   
75%      32.00000   15.000000   148.500000   453.600000    1018.440000   
max    1899.00000   15.000000  2278.800000  7360.000000   17470.800000   

       LastEditedBy  
count    461.000000  
mean      11.080260  
std        5.358461  
min        2.000000  
25%        6.000000  
50%       11.000000  
75%       16.000000  
max       20.000000

### **Question 10**
Perform the action that you recommended above. How many rows are left in the data frame?

Group of answer choices
* 230,087 (answer if I drop them)
* 230,548 (answer if I don't drop them)
* ~~230,100~~
* ~~230,542~~
### Answer: 230,087

In [9]:
### making a dataframe for the scenario of keeping rows
df_alt = df.copy()

### dropping rows with Null in both StockItemID and Description.
df = df.dropna(subset = ['StockItemID', 'Description'], how = 'all')

### Exploring both scenarios from this point on...
print('Rows if I drop them:')
print(df.shape[0])

print('\nRows if I keep them:')
print(df_alt.shape[0])

Rows if I drop them:
230087

Rows if I keep them:
230548


### **Question 11**
Observe the TaxRate column, which contains some negative values. Based on your domain knowledge and best guess of what tax rates should be, which of the following actions might you recommend for preparing this data set for analysis?

Group of answer choices
* turn the negative tax rates into positive tax rates by multiplying by -1
* change the negative tax rates to 0
* leave the negative tax rates as they are
* change all tax rates to be negative
### Answer: turn the negative tax rates into positive tax rates by multiplying by -1
* I'll be using abs() on all rows instead of filtering for negative values to multiply by -1

In [10]:
print('\nDrop Scenario:')
display(df['TaxRate'].describe())

print('\n\nKeep Scenario:')
display(df_alt['TaxRate'].describe())


Drop Scenario:


count    230087.000000
mean         14.463268
std           3.904501
min         -15.000000
25%          15.000000
50%          15.000000
75%          15.000000
max          15.000000
Name: TaxRate, dtype: float64



Keep Scenario:


count    230548.000000
mean         14.463409
std           3.904056
min         -15.000000
25%          15.000000
50%          15.000000
75%          15.000000
max          15.000000
Name: TaxRate, dtype: float64

### **Question 12**
Perform the action that you recommended above. What is the average TaxRate?

Group of answer choices
* 14.06
* 14.98
* 14.76
* 14.63
### Answer: 14.98 (in either scenario when rounded)

In [11]:
print('\nAverage TaxRate - Drop Scenario:')
df['TaxRate'] = abs(df['TaxRate'])
print(round(df['TaxRate'].mean(),2))

print('\nAverage TaxRate - Keep Scenario:')
df_alt['TaxRate'] = abs(df_alt['TaxRate'])
print(round(df_alt['TaxRate'].mean(),2))


Average TaxRate - Drop Scenario:
14.98

Average TaxRate - Keep Scenario:
14.98


### **Question 13**
Observe the column TaxAmount, which indicates the amount of tax paid for each invoice line. The TaxAmount can be calculated using the following formula:
```
TaxAmount = (UnitPrice * Quantity * TaxRate) / 100
```

Based on this knowledge, the number of null values in the data set, and the questions being asked by the executives, what do you recommend doing to remove null values from this column?

Group of answer choices
* impute the missing values
* do nothing
* drop the TaxAmount column
* drop the rows with null values in the TaxAmount column
### Answer: impute the missing values

In [12]:
print('\nNulls in TaxAmount - Drop Scenario:')
print(df['TaxAmount'].isna().sum())

print('\nNulls in TaxAmount - Keep Scenario:')
print(df_alt['TaxAmount'].isna().sum())


Nulls in TaxAmount - Drop Scenario:
120679

Nulls in TaxAmount - Keep Scenario:
120925


### **Question 14**
Perform the action that you recommended above. What is the average TaxAmount after removing the null values?

Group of answer choices
* 158.91
* 112.95
* 110.42
* 113.29
### Answer: 112.95
* The 'Drop' scenario average is the only calculated value found in the group of answer choices (112.95).
* "Drop the rows with null values in the Description and StockItemID columns" must have been the expected answer for Question 9.
* My selected answers for this project will come from the 'Drop Scenario.'

In [13]:
df.loc[(df['TaxAmount'].isna()),'TaxAmount'] = (df['UnitPrice'] * df['Quantity'] * df['TaxRate']) / 100     # (UnitPrice * Quantity * TaxRate) / 100

print('\nAverage TaxAmount - Drop Scenario:')
print(round(df['TaxAmount'].mean(),2))

# Keep Scenario
df_alt.loc[(df_alt['TaxAmount'].isna()),'TaxAmount'] = (df_alt['UnitPrice'] * df_alt['Quantity'] * df_alt['TaxRate']) / 100     # (UnitPrice * Quantity * TaxRate) / 100

print('\nAverage TaxAmount - Keep Scenario:')
print(round(df_alt['TaxAmount'].mean(),2))


Average TaxAmount - Drop Scenario:
112.95

Average TaxAmount - Keep Scenario:
112.98


### **Question 15**
Observe the column ExtendedPrice, which indicates the total price paid by the customer. The ExtendedPrice can be calculated using the following formula:
```
ExtendedPrice = (UnitPrice * Quantity) + TaxAmount
```
Based on this knowledge, the number of null values in the data set, and the questions being asked by the executives, what do you recommend doing to remove null values from this column?

Group of answer choices
* impute the missing values
* drop the column ExtendedPrice
* do nothing
* drop the rows with null values in the ExtendedPrice column
### Answer: impute the missing values

In [14]:
print('\nNulls in ExtendedPrice - Drop Scenario:')
print(df['ExtendedPrice'].isna().sum())

print('\nNulls in ExtendedPrice - Keep Scenario:')
print(df_alt['ExtendedPrice'].isna().sum())


Nulls in ExtendedPrice - Drop Scenario:
117229

Nulls in ExtendedPrice - Keep Scenario:
117467


### **Question 16**
Perform the action that you recommended above. Treat the values of -999 as an input error and perform the recommended action on that value as well. After addressing these issues, what is the average ExtendedPrice?

Group of answer choices
* 1,878.01
* 867.64
* 552.08
* 809.99
### Answer: 867.64

In [15]:
df.loc[(df['ExtendedPrice'].isna()) | (df['ExtendedPrice'] == -999),'ExtendedPrice'] = (df['UnitPrice'] * df['Quantity']) + df['TaxAmount']     # ExtendedPrice = (UnitPrice * Quantity) + TaxAmount

print('\nAverage ExtendedPrice - Drop Scenario:')
print(round(df['ExtendedPrice'].mean(),2))


df_alt.loc[(df_alt['ExtendedPrice'].isna()) | (df_alt['ExtendedPrice'] == -999),'ExtendedPrice'] = (df_alt['UnitPrice'] * df_alt['Quantity']) + df_alt['TaxAmount']     # ExtendedPrice = (UnitPrice * Quantity) + TaxAmount

print('\nAverage ExtendedPrice - Keep Scenario:')
print(round(df_alt['ExtendedPrice'].mean(),2))


Average ExtendedPrice - Drop Scenario:
867.64

Average ExtendedPrice - Keep Scenario:
867.83


### **Question 17**
Observe values in the Quantity and ExtendedPrice columns. Use the z-score method with a threshold of 3 to determine which rows are outliers in these columns. Make sure that you get rows that have a z-score greater than or equal to 3.
Based on the Quantity column, how many rows could be considered outliers?

Group of answer choices
* 5,987
* 2,667
* 1,887
* 7,006
### Answer: 5,987

In [16]:
### Create Z-Score columns
df['Z-Score_Price'] = (df['ExtendedPrice'] - df['ExtendedPrice'].mean()) / df['ExtendedPrice'].std()
df['Z-Score_Qty'] = (df['Quantity'] - df['Quantity'].mean()) / df['Quantity'].std()

# How many >= 3
print('\nQty outliers - Drop Scenario:')
print(len(df[df['Z-Score_Qty'] >= 3]))

### Create Z-Score columns - Keep Scenario
df_alt['Z-Score_Price'] = (df_alt['ExtendedPrice'] - df_alt['ExtendedPrice'].mean()) / df_alt['ExtendedPrice'].std()
df_alt['Z-Score_Qty'] = (df_alt['Quantity'] - df_alt['Quantity'].mean()) / df_alt['Quantity'].std()

# How many >= 3
print('\nQty outliers - Keep Scenario:')
print(len(df_alt[df_alt['Z-Score_Qty'] >= 3]))


Qty outliers - Drop Scenario:
5987

Qty outliers - Keep Scenario:
6002


### **Question 18**
Observe values in the Quantity and ExtendedPrice columns. Use the z-score method with a threshold of 3 to determine which rows are outliers in these columns. Make sure that you get rows that have a z-score greater than or equal to 3.

Based on the ExtendedPrice column, how many rows could be considered outliers?

Group of answer choices
* 4,522
* 7,113
* 6,667
* 1,546
### Answer: 4,522

In [17]:
print('Price Outliers - Drop Scenario')
# How many >= 3
display(len(df[(df['Z-Score_Price'] >= 3)]))

print('\nPrice Outliers - Keep Scenario')
# How many >= 3
display(len(df_alt[(df_alt['Z-Score_Price'] >= 3)]))

Price Outliers - Drop Scenario


4522


Price Outliers - Keep Scenario


4535

### **Question 19**
Based on your understanding of outliers and the answer above, choose the most appropriate action to handle the outliers in the ExtendedPrice column.

Group of answer choices
* use imputation to fill in the values again
* remove the outliers from the data set
* divide the outlier values by 10 to make them smaller and not outliers
* make a note of the outliers in the analysis
### Answer: make a note of the outliers in the analysis

### **Question 20**
The data set provided should have a unique ID for each invoice line, but there can be repeated invoice IDs. This is because several unique invoice lines can share the same invoice ID. Given this information, how would you approach checking for duplicate rows in the data set?

Group of answer choices
* check for duplicates using the InvoiceLineID column only
* check for duplicates using both the InvoiceID and InvoiceLineID columns
* check for duplicates using the entire row of data
* check for duplicates using the InvoiceID column only
### Answer: check for duplicates using the InvoiceLineID column only
This one works, too:
* check for duplicates using both the InvoiceID and InvoiceLineID columns

In [18]:
print('Drop Scenario')
print('\nDuplicates:')
print(len(df.loc[df.duplicated('InvoiceLineID')]))
print('How many would be left:')
print(len(df.drop_duplicates('InvoiceLineID')))

print('\n\nKeep Scenario')
print('\nDuplicates:')
print(len(df_alt.loc[df_alt.duplicated('InvoiceLineID')]))
print('How many would be left:')
print(len(df_alt.drop_duplicates('InvoiceLineID')))

Drop Scenario

Duplicates:
2276
How many would be left:
227811


Keep Scenario

Duplicates:
2283
How many would be left:
228265


### **Question 21**
Perform the action that you recommended above and drop duplicate rows from the data set. How many total rows are in the data set after performing this action?

Group of answer choices
* 203,803
* 199,007
* 227,811
* 201,989
### Answer: 227,811

In [19]:
df.drop_duplicates(subset=['InvoiceID', 'InvoiceLineID'], inplace = True)
print('Rows remaining - Drop Scenario:')
print(df.shape[0])

df_alt.drop_duplicates(subset=['InvoiceID', 'InvoiceLineID'], inplace = True)
print('\nRows remaining - Keep Scenario:')
print(df_alt.shape[0])

Rows remaining - Drop Scenario:
227811

Rows remaining - Keep Scenario:
228265


### **Question 22**
At this point, you should have a data set that is mostly free of null values (only has some null values in the Description column). Use this data set to answer the questions below.

Which item has the highest total quantity sold? Use the Description and Quantity columns.

Group of answer choices
* Void fill 200 L bag (White) 200L
* "The Gu" red shirt XML tag t-shirt (Black) 3XL
* White chocolate snow balls 250g
* Black and orange fragile despatch tape 48mmx75m
### Answer: Black and orange fragile despatch tape 48mmx75m

In [20]:
print('\n5 most sold items by total quantity:\n')
display(
  df[['Description', 'Quantity']]
    .groupby('Description').sum()
    .sort_values(by = 'Quantity', ascending = False).reset_index()
    .rename(columns={"Quantity" : "TotalQty"})
    .head()
       )


5 most sold items by total quantity:



Description  TotalQty
0   Black and orange fragile despatch tape 48mmx75m    207108
1  Black and orange fragile despatch tape 48mmx100m    193572
2                     Clear packaging tape 48mmx75m    158392
3        3 kg Courier post bag (White) 300x190x95mm    154525
4             Shipping carton (Brown) 356x356x279mm    151525

### **Question 23**
Which item has generated the most total profit? Use the Description and LineProfit columns.

Group of answer choices
* "The Gu" red shirt XML tag t-shirt (Black) 3XL
* 20 mm Double sided bubble wrap 50m
* Halloween zombie mask (Light Brown) XL
* Air cushion machine (Blue)
### Answer: 20 mm Double sided bubble wrap 50m

In [21]:
print('\nThe top 5 items with most total profit:\n')
display(
  df[['Description', 'LineProfit']]
    .groupby(by = 'Description').sum()
    .sort_values(by = 'LineProfit', ascending = False).reset_index()
    .rename(columns={"LineProfit" : "TotalProfit"})
    .head()
       )


The top 5 items with most total profit:



Description  TotalProfit
0        20 mm Double sided bubble wrap 50m    5283560.0
1                Air cushion machine (Blue)    4418898.0
2  32 mm Anti static bubble wrap (Blue) 50m    3514220.0
3  10 mm Anti static bubble wrap (Blue) 50m    3439800.0
4        32 mm Double sided bubble wrap 50m    2915530.0

### **Question 24**
Which item is purchased in greatest quantities in a single order, on average? Use the Quantity and Description columns.

Group of answer choices
* Superhero action jacket (Blue) XXL
* Black and orange fragile despatch tape 48mmx75m
* DBA joke mug - SELECT caffeine FROM mug (Black)
* Shipping carton (Brown) 356x356x279mm
### Answer: Black and orange fragile despatch tape 48mmx75m

In [22]:
print('\nThe top 5 items purchased in the greatest quantities in a single order, on average:\n')
display(
  df[['Description', 'Quantity']]
    .groupby(by = 'Description').mean()
    .sort_values(by = 'Quantity', ascending = False).reset_index()
    .rename(columns={"Quantity" : "AverageQty"})
    .head()
        )


The top 5 items purchased in the greatest quantities in a single order, on average:



Description  AverageQty
0   Black and orange fragile despatch tape 48mmx75m  199.526012
1  Black and orange fragile despatch tape 48mmx100m  199.148148
2                     Clear packaging tape 48mmx75m  145.313761
3        3 kg Courier post bag (White) 300x190x95mm  144.012116
4             Shipping carton (Brown) 356x356x279mm  141.612150